src/sentry/sentry_metrics/indexer/indexer_consumer.py

```diff
@@ -0,0 +1,57 @@
+import logging
+from typing import Any, Dict
+
+from django.conf import settings
+
+from sentry.sentry_metrics import indexer
+from sentry.sentry_metrics.indexer.base import UseCase
+from sentry.utils import json, kafka_config
+from sentry.utils.batching_kafka_consumer import AbstractBatchWorker
+from sentry.utils.kafka import create_batching_kafka_consumer
+from sentry.utils.pubsub import KafkaPublisher
+
+logger = logging.getLogger(__name__)
+
+snuba_metrics = settings.KAFKA_TOPICS[settings.KAFKA_SNUBA_METRICS]
+
+
+def get_metrics_consumer(topic=None, **options) -> None:
+    return create_batching_kafka_consumer(
+        {"ingest-metrics"}, worker=MetricsIndexerWorker(), **options
+    )
+
+
+class MetricsIndexerWorker(AbstractBatchWorker):
+    def process_message(self, message):
+        original: Dict[str, Any] = json.loads(message.value())
+        new_message = original.copy()
+
+        org_id = int(new_message["org_id"])
+        metric_id = indexer.record(org_id, UseCase.METRIC, new_message["name"])
+
+        tags = new_message["tags"]
+        new_tags = {}
+        for tag_k, tag_v in tags.items():
+            new_k = indexer.record(org_id, UseCase.TAG_KEY, tag_k)
+            new_v = indexer.record(org_id, UseCase.TAG_VALUE, tag_v)
+            new_tags[new_k] = new_v
+
+        new_message["tags"] = new_tags
+        new_message["metric_id"] = metric_id
+        new_message["retention_days"] = 90
+        return new_message
+
+    def flush_batch(self, batch):
+        # produce the translated message to snuba-metrics topic
+        global snuba_metrics
+        for message in batch:
+            cluster_name = snuba_metrics["cluster"]
+            snuba_metrics_publisher = KafkaPublisher(
+                kafka_config.get_kafka_producer_cluster_options(cluster_name),
+                asynchronous=False,
+            )
+            snuba_metrics_publisher.publish(snuba_metrics["topic"], json.dumps(message))
```

Comment:
This would flush the messages for every message, which is not ideal considering you have batches of messages and you can flush before committing on the ingest topic only once per batch of messages.
We should also set the callback for each message.

Reply:
Some follow up on what a consumer should do:

When writing a consumer like this that needs to achieve a high throughput, there are a few elements to take into account and a few requirements:
- We cannot auto commit, because we would be committing before processing the message. If there is an error in processing the message is lost. This is taken into account as we do not auto commit
- We don't need to commit after every offset. Less load on the broker and network
- We must not loose messages. So if an error happens at any point during processing we cannot commit the entire batch but only the portion that we sent to Kafka
- producing is asynchronous. So we need to wait for the callback before being sure that the message is persisted in a kafka topic https://docs.confluent.io/clients-confluent-kafka-python/current/overview.html#asynchronous-writes
- We should really avoid duplicates, as they would not be deduplicated in Clickhouse since metrics are stored in a pre-aggregated way. So if there is an error during processing that causes the consumer to crash we should have committed up to the last acknowledged message (acknowledged meaning that we did receive the callback from the producer).
- We should keep producing asynchronously and not flush every message individually as that would have a real impact on throughput.
- Exactly one semantics is technically not achievable as nobody can deduplicate messages. But we can make duplication extremely rare (basically only in case Kafka commit fails multiple times or the consumer crashes for out of memory without being able to commit and after flushing.

So there are a few ways to do that:
- Simple batching consumer. Do the processing phase, then the batch flush sends all the messages, at the end flushes and waits for callbacks. It only commits on kafka the offset of the last callback received.
- Something like the cdc link above. Keep producing messages as soon as they are processed and periodically commit the last offset we got the callback for.